### PR TITLE
[dynamo] Don't copy the graph during checkpointing (copy_graphstate)

### PR DESCRIPTION
**Pull request by mlazos**
copy_graphstate is called a ton, this makes copy_graphstate a lot faster, helps with https://github.com/pytorch/torchdynamo/issues/1803

tag each graph node with a timestamp, when checkpointing store the timestamp, when restoring remove nodes older than the timestamp stored in the state. This essentially has the same behavior as the original impl, just doesn't copy the whole graph.

cc @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/_dynamo/output_graph.py [hunk 1]
```diff
@@ -567,3 +574,16 @@ def create_proxy(
         rv.node.stack_trace = nn_module_stack_str + " | ".join(msgs)
 
         return rv
+
+    def create_node(self, *args, **kwargs):
+        node = super().create_node(*args, **kwargs)
+        node.timestamp = self.timestamp
```
Comment (jansel) [inline torch/_dynamo/output_graph.py:580]: put this in `node.meta["creation_timestamp"]`

File: torch/_dynamo/output_graph.py [hunk 2]
```diff
@@ -114,6 +118,10 @@ def __init__(
         # all current placeholder node names
         self.name_to_input = collections.OrderedDict()
 
+        # Used to avoid iterating over the linked list of graph nodes (slow)
+        # the keys contain the nodes, all values are None
+        self.nodes = collections.OrderedDict()
```
Comment (jansel) [inline torch/_dynamo/output_graph.py:123]: This adds complexity.  What is the perf difference of this versus just the timestamp change?

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_dynamo/output_graph.py (hunk 2 above)
Comment (mlazos) [reply torch/_dynamo/output_graph.py:123]: Yeah this was more experimental, I can probably remove this if we no longer copy the graph, `restore_checkpoint` is obviously called a lot less.

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**